packages/core/src/types/shapeEvalTypes.ts

```diff
@@ -0,0 +1,477 @@
+import { VarAD } from "./adTypes";
+import { ASTNode, Identifier } from "./ASTTypes";
+import { StyleError } from "./errors";
+import { SubPath, Value } from "./shapeTypes";
+
+/**
+ * The input parameters to computations/objectives/constraints in Style. It can be either an entire shape (`IGPI`) or a value (`IVal`).
+ */
+export type ArgVal<T> = IGPI<T> | IVal<T>;
+
+export interface IGPI<T> {
+  tag: "GPI";
+  contents: GPI<T>;
+}
+
+/**
+ * A shape (Graphical Primitive Instance, aka GPI) in penrose has a type (_e.g._ `Circle`) and a set of properties (_e.g._ `center`). Each property this a value tagged with its type.
+ */
+export type GPI<T> = [string, { [k: string]: Value<T> }];
+
+export interface IVal<T> {
+  tag: "Val";
+  contents: Value<T>;
+}
+
+export type Field = string;
+export type Name = string;
+export type Property = string;
+export type FExpr = FieldExpr<VarAD>;
+export type ShapeTypeStr = string;
+export type PropID = string;
+export type GPIMap = { [k: string]: TagExpr<VarAD> };
+export type FieldDict = { [k: string]: FieldExpr<VarAD> };
+
+export type StyleOptFn = [string, Expr[]]; // Objective or constraint
+
+// NOTE: To make a deep clone, use `clone` from `rfdc`
+/**
+ * Translation represents the computational graph compiled from a trio of Penrose programs.
+ */
+export type Translation = ITrans<VarAD>;
+
+export type Trans = TrMap<VarAD>;
+
+export type TrMap<T> = { [k: string]: { [k: string]: FieldExpr<T> } };
+
+export interface ITrans<T> {
+  // TODO: compGraph
+  trMap: TrMap<T>;
+  warnings: StyleError[];
+}
+
+export type FieldExpr<T> = IFExpr<T> | IFGPI<T>;
+
+export interface IFExpr<T> {
+  tag: "FExpr";
+  contents: TagExpr<T>;
+}
+
+export interface IFGPI<T> {
+  tag: "FGPI";
+  contents: GPIExpr<T>;
+}
+
+export type GPIProps<T> = { [k: string]: TagExpr<T> };
+
+export type GPIExpr<T> = [string, { [k: string]: TagExpr<T> }];
+export type TagExpr<T> = IOptEval<T> | IDone<T> | IPending<T>;
+
+export interface IOptEval<T> {
+  tag: "OptEval";
+  contents: Expr;
+}
+
+export interface IDone<T> {
+  tag: "Done";
+  contents: Value<T>;
+}
+
+export interface IPending<T> {
+  tag: "Pending";
+  contents: Value<T>;
+}
+
+export type Expr =
```

Comment:
This belongs to style AST technically

Reply:
the whole thing? all the sum type members r in there